packages/webpack-cli/src/plugins/dotenv-webpack-plugin.ts

```diff
@@ -0,0 +1,253 @@
+import dotenv from "dotenv";
+import { Compiler, DefinePlugin } from "webpack";
+
+interface EnvVariables {
+  [key: string]: string;
+}
+
+interface Logger {
+  info(message: string): void;
+  warn(message: string): void;
+  error(message: string): void;
+  log(message: string): void;
+}
+
+interface DotenvConfig {
+  paths?: string[];
+  prefixes?: string[];
+  systemvars?: boolean;
+  allowEmptyValues?: boolean;
+  expand?: boolean;
+  ignoreStub?: boolean;
+  safe?: boolean;
+}
+
+const interpolate = (env: string, vars: EnvVariables): string => {
+  const matches = env.match(/\$([a-zA-Z0-9_]+)|\${([a-zA-Z0-9_]+)}/g) || [];
+  matches.forEach((match) => {
+    env = env.replace(match, interpolate(vars[match.replace(/\$|{|}/g, "")] || "", vars));
+  });
+  return env;
+};
+
+declare interface InputFileSystem {
+  readFile: (
+    arg0: string,
+    arg1: (arg0?: null | NodeJS.ErrnoException, arg1?: string | Buffer) => void,
+  ) => void;
+}
+
+const isMainThreadElectron = (target: string | undefined): boolean =>
+  !!target && target.startsWith("electron") && target.endsWith("main");
+
+export class Dotenv {
+  #logger!: Logger;
+  #options: DotenvConfig;
+  #inputFileSystem!: InputFileSystem;
+
+  constructor(config: DotenvConfig = {}) {
+    this.#options = {
+      paths: process.env.NODE_ENV
+        ? [
+            ".env",
+            ...(process.env.NODE_ENV === "test" ? [] : [".env.local"]),
+            `.env.${process.env.NODE_ENV}`,
+            `.env.${process.env.NODE_ENV}.local`,
+          ]
+        : [],
+      prefixes: ["process.env.", "import.meta.env."],
+      allowEmptyValues: true,
+      expand: true,
+      ...config,
+    };
+  }
+
+  public apply(compiler: Compiler): void {
+    this.#inputFileSystem = compiler.inputFileSystem;
+    this.#logger = compiler.getInfrastructureLogger("dotenv-webpack-plugin");
+    if (!this.#options.paths) {
+      this.#options.paths = [
+        ".env",
+        ".env.local",
+        `.env.${compiler.options.mode}`,
+        `.env.${compiler.options.mode}.local`,
+      ];
+    }
```

Comment:
This is invalid, please use in constructor:
```
[
  ".env",
  ...(process.env.NODE_ENV === "test" ? [] : [".env.local"]),
  `.env.[mode]`,
  `.env.[mode].local`,
]
```

then in `for (const path of paths) {` just replace `[mode]` on `process.env.NODE_ENV || compiler.options.mode`